cmd/internal/http/handlers.go

```diff
@@ -92,31 +93,41 @@ func RegisterHandlers(mux httpmux.Mux, containerManager manager.Manager, httpAut
 	return nil
 }
 
-// RegisterPrometheusHandler creates a new PrometheusCollector and configures
+// RegisterPrometheusHandler creates a new ContainerCollector and configures
 // the provided HTTP mux to handle the given Prometheus endpoint.
 func RegisterPrometheusHandler(mux httpmux.Mux, resourceManager manager.Manager, prometheusEndpoint string,
 	f metrics.ContainerLabelsFunc, includedMetrics container.MetricSet) {
-	goCollector := prometheus.NewGoCollector()
-	processCollector := prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{})
-	machineCollector := metrics.NewPrometheusMachineCollector(resourceManager, includedMetrics)
+
+	r := prometheus.NewRegistry()
+	r.MustRegister(
+		prometheus.NewGoCollector(),
+		prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{}),
+	)
+	container := metrics.NewContainerCollector(resourceManager, f, includedMetrics, clock.RealClock{})
+	machine := metrics.NewMachineCollector(resourceManager, includedMetrics)
+
+	nameTypeCache := metrics.NewCachedGatherer(container.Collect, machine.Collect)
+	nameDockerCache := metrics.NewCachedGatherer(container.Collect, machine.Collect)
+	nameTypeGatherer := prometheus.NewMultiTRegistry(prometheus.ToTransactionalGatherer(r), nameTypeCache)
+	nameDockerGatherer := prometheus.NewMultiTRegistry(prometheus.ToTransactionalGatherer(r), nameDockerCache)
 
 	mux.Handle(prometheusEndpoint, http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		// TODO(bwplotka): Why we ask for full object if we override half of fields?
 		opts, err := api.GetRequestOptions(req)
 		if err != nil {
 			http.Error(w, "No metrics gathered, last error:\n\n"+err.Error(), http.StatusInternalServerError)
 			return
 		}
-		opts.Count = 1        // we only want the latest datapoint
-		opts.Recursive = true // get all child containers
-
-		r := prometheus.NewRegistry()
-		r.MustRegister(
-			metrics.NewPrometheusCollector(resourceManager, f, includedMetrics, clock.RealClock{}, opts),
-			machineCollector,
-			goCollector,
-			processCollector,
-		)
-		promhttp.HandlerFor(r, promhttp.HandlerOpts{ErrorHandling: promhttp.ContinueOnError}).ServeHTTP(w, req)
+		opts.Count = 1        // We only want the latest datapoint.
+		opts.Recursive = true // Get all child containers.
```

Comment:
These options don't appear to be provided to the nameTypeGatherer anymore; previously we'd plug these in via the NewPrometheusCollector.

Reply:
got it now, you mean this 

![image](https://user-images.githubusercontent.com/6950331/154316676-b3d12ba5-6f8d-4676-a984-dcc57ffca1ad.png)

Reply:
You comment was on point, we DO forget using `UpdateOnMaxAge` here (and in our k8s impl). My bad, fixed (: